service/deferrer/service.go

```diff
@@ -31,3 +41,81 @@ func New(config Config) (*Service, error) {
 
 	return s, nil
 }
+
+// ShouldDefer finds corresponding DrainerConfig for the POD it's running in
+// and checks if node in current POD is drained yet. If DrainerConfig doesn't
+// exist or doesn't have Drained or Timeout condition, node termination should
+// be deferred.
+//
+// Current POD name and namespace are picked from environment variables with
+// corresponding keys defined in constants EnvKeyMyPodName &
+// EnvKeyMyPodNamespace. Defining these env variables is most conveniently
+// achieved by utilizing Kubernetes Downward API:
+// https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/
+func (s *Service) ShouldDefer(ctx context.Context) (bool, error) {
+	s.logger.LogCtx(ctx, "level", "debug", "message", "finding out if node termination must be deferred")
+
+	podName, err := s.getPodName()
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	podNamespace, err := s.getPodNamespace()
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	s.logger.LogCtx(ctx, "level", "debug", "message", "finding drainerconfig for pod")
+
+	drainerConfig, err := s.g8sClient.CoreV1alpha1().DrainerConfigs(podNamespace).Get(podName, metasv1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		s.logger.LogCtx(ctx, "level", "debug", "message", "did not find drainerconfig")
+		s.logger.LogCtx(ctx, "level", "debug", "message", "node is not drained yet")
+		s.logger.LogCtx(ctx, "level", "debug", "message", "node termination must be deferred")
+		return true, nil
+	} else if err != nil {
+		return true, microerror.Mask(err)
+	}
+
+	s.logger.LogCtx(ctx, "level", "debug", "message", "found drainerconfig for pod")
+
+	hasCondition := false
+
+	if drainerConfig.Status.HasDrainedCondition() {
+		s.logger.LogCtx(ctx, "level", "debug", "message", "pod drainerconfig has drained condition")
+		hasCondition = true
+	}
+
+	if drainerConfig.Status.HasTimeoutCondition() {
+		s.logger.LogCtx(ctx, "level", "debug", "message", "pod drainerconfig has timeout condition")
+		hasCondition = true
+	}
+
+	if hasCondition {
+		s.logger.LogCtx(ctx, "level", "debug", "message", "node termination don't need to be deferred")
+		return false, nil
+	}
+
+	s.logger.LogCtx(ctx, "level", "debug", "message", "node is not drained yet")
+	s.logger.LogCtx(ctx, "level", "debug", "message", "node termination must be deferred")
+
+	return true, nil
+}
+
+func (s *Service) getPodName() (string, error) {
+	podName := os.Getenv(EnvKeyMyPodName)
+	if podName == "" {
+		return "", microerror.Maskf(invalidConfigError, "pod name not present in runtime")
```

Comment:
WDYT about printing env var name in the message so it's easy to fix it w/o looking into the code?

Reply:
Sure, fix on its way :slightly_smiling_face: